From 89e52580c4b3f1d095d7966c8cf8607c4b39f58c Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Fri, 29 Dec 2017 16:43:59 +0800
Subject: [PATCH 3164/5242] MLK-17312-4 usb: cdns3: gadget: fix the
 recognition issue when connection before load
 module

commit  397410457133b5a83a7bd565784cfab8d179fb5c from
https://source.codeaurora.org/external/imx/linux-imx.git

At imx8qm/imx8qxp A0 chip, there is a vbus toggle issue, so we need to
force the vbus as high before connection, otherwise, there will be endless
connect/disconnect interrupts for USB2 and causes enumeration failure.
The current work flow only cover this during the role switch, but omit
it when the connection has established at module probe routine.

This patch fixes it by moving force vbus operation to cdns_set_role to
cover both static and dynamic recognition issue.

Acked-by: Jun Li <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/cdns3/core.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/usb/cdns3/core.c b/drivers/usb/cdns3/core.c
index 9ba602a..8b0f5da 100644
--- a/drivers/usb/cdns3/core.c
+++ b/drivers/usb/cdns3/core.c
@@ -159,6 +159,8 @@ static void cdns_set_role(struct cdns3 *cdns, enum cdns3_roles role)
 		writel(value, cdns->none_core_regs + USB3_CORE_CTRL1);
 		mdelay(1);
 		cdns3_usb_phy_init(cdns->phy_regs);
+		/* Force B Session Valid as 1 */
+		writel(0x0060, cdns->phy_regs + 0x380a4);
 		mdelay(1);
 
 		value = readl(cdns->none_core_regs + USB3_INT_REG);
@@ -192,6 +194,8 @@ static void cdns_set_role(struct cdns3 *cdns, enum cdns3_roles role)
 		writel(value, cdns->none_core_regs + USB3_CORE_CTRL1);
 
 		cdns3_usb_phy_init(cdns->phy_regs);
+		/* Force B Session Valid as 1 */
+		writel(0x0060, cdns->phy_regs + 0x380a4);
 		value = readl(cdns->none_core_regs + USB3_INT_REG);
 		value |= DEV_INT_EN;
 		writel(value, cdns->none_core_regs + USB3_INT_REG);
@@ -398,10 +402,6 @@ static int cdns3_do_role_switch(struct cdns3 *cdns, enum cdns3_roles role)
 	}
 
 	cdns_set_role(cdns, role);
-	if (role == CDNS3_ROLE_GADGET || role == CDNS3_ROLE_HOST)
-		/* Force B Session Valid as 1 */
-		writel(0x0060, cdns->phy_regs + 0x380a4);
-
 	ret = cdns3_role_start(cdns, role);
 	if (ret) {
 		/* Back to current role */
-- 
1.7.9.5

